# Challenge Telco Communication

Ho a disposizione un dataset con le caratteristiche dei clienti, voglio costruire un classificatore che prevede se un cliente resterà tale o meno. La fetures target nel dataset è il Churn,che contiene Yes o No in base alle caratteristiche del cliente che sono tutte le altre features del dataset.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split #per dividere il dataset in dati di training e dati di test 
from sklearn.tree import DecisionTreeClassifier #per usare un classificatore Decision Tree
from sklearn.metrics import accuracy_score  #per usare come metrica l'accuratezza
from sklearn.metrics import confusion_matrix #per usare una matrice di confusione per misurare le prestazioni 
from sklearn.preprocessing import LabelEncoder #per trasformare i tipi in int
from sklearn.metrics import classification_report

In [2]:
dataset = pd.read_csv('../Challenge2/WA_Fn-UseC_-Telco-Customer-Churn.train.csv') #importo i dati 

Importo il dataset che si trova nella cartella Challenge2 

In [3]:
dataset.head()

Churn  customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0    No  4223-BKEOR  Female              0      No        Yes      21   
1    No  6035-RIIOM  Female              0      No         No      54   
2   Yes  3797-VTIDR    Male              0     Yes         No       1   
3   Yes  2568-BRGYX    Male              0      No         No       4   
4    No  2775-SEFEE    Male              0      No        Yes       0   

  PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0          Yes                No             DSL  ...           No   
1          Yes               Yes     Fiber optic  ...          Yes   
2           No  No phone service             DSL  ...           No   
3          Yes                No     Fiber optic  ...           No   
4          Yes               Yes             DSL  ...          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0              Yes          No          No             Yes        One year   
1               No          No         Yes             Yes        Two year   
2               No          No          No              No  Month-to-month   
3               No          No          No              No  Month-to-month   
4               No         Yes          No              No        Two year   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  
0               No               Mailed check          64.85        1336.8  
1              Yes  Bank transfer (automatic)          97.20       5129.45  
2              Yes           Electronic check          23.45         23.45  
3              Yes           Electronic check          70.20        237.95  
4              Yes  Bank transfer (automatic)          61.90                

[5 rows x 21 columns]

Visualizzo le prime 5 righe del dataset

In [4]:
dataset.shape #ci sono 5634 righe e 21 colonne

(5634, 21)

Utilizzo questo metodo per controllare se ci sono dei valori nulli nelle righe del dataset 

In [5]:
dataset.isna().any() #non ci sono righe nulle 

Churn               False
customerID          False
gender              False
SeniorCitizen       False
Partner             False
Dependents          False
tenure              False
PhoneService        False
MultipleLines       False
InternetService     False
OnlineSecurity      False
OnlineBackup        False
DeviceProtection    False
TechSupport         False
StreamingTV         False
StreamingMovies     False
Contract            False
PaperlessBilling    False
PaymentMethod       False
MonthlyCharges      False
TotalCharges        False
dtype: bool

In [6]:
features_categoriche = dataset.select_dtypes(include = 'object') #selectiono le features che sono object
features_categoriche

Churn  customerID  gender Partner Dependents PhoneService  \
0       No  4223-BKEOR  Female      No        Yes          Yes   
1       No  6035-RIIOM  Female      No         No          Yes   
2      Yes  3797-VTIDR    Male     Yes         No           No   
3      Yes  2568-BRGYX    Male      No         No          Yes   
4       No  2775-SEFEE    Male      No        Yes          Yes   
...    ...         ...     ...     ...        ...          ...   
5629   Yes  0684-AOSIH    Male     Yes         No          Yes   
5630    No  5982-PSMKW  Female     Yes        Yes          Yes   
5631    No  8044-BGWPI    Male     Yes        Yes          Yes   
5632   Yes  7450-NWRTR    Male      No         No          Yes   
5633    No  4795-UXVCJ    Male      No         No          Yes   

         MultipleLines InternetService       OnlineSecurity  \
0                   No             DSL                  Yes   
1                  Yes     Fiber optic                   No   
2     No phone service             DSL                   No   
3                   No     Fiber optic                   No   
4                  Yes             DSL                  Yes   
...                ...             ...                  ...   
5629                No     Fiber optic                  Yes   
5630               Yes             DSL                  Yes   
5631                No              No  No internet service   
5632               Yes     Fiber optic                   No   
5633                No              No  No internet service   

             OnlineBackup     DeviceProtection          TechSupport  \
0                      No                  Yes                   No   
1                     Yes                   No                   No   
2                      No                   No                   No   
3                      No                   No                   No   
4                     Yes                   No                  Yes   
...                   ...                  ...                  ...   
5629                   No                   No                   No   
5630                  Yes                  Yes                  Yes   
5631  No internet service  No internet service  No internet service   
5632                   No                  Yes                   No   
5633  No internet service  No internet service  No internet service   

              StreamingTV      StreamingMovies        Contract  \
0                      No                  Yes        One year   
1                     Yes                  Yes        Two year   
2                      No                   No  Month-to-month   
3                      No                   No  Month-to-month   
4                      No                   No        Two year   
...                   ...                  ...             ...   
5629                  Yes                  Yes  Month-to-month   
5630                  Yes                  Yes        Two year   
5631  No internet service  No internet service  Month-to-month   
5632                  Yes                  Yes  Month-to-month   
5633  No internet service  No internet service        One year   

     PaperlessBilling              PaymentMethod TotalCharges  
0                  No               Mailed check       1336.8  
1                 Yes  Bank transfer (automatic)      5129.45  
2                 Yes           Electronic check        23.45  
3                 Yes           Electronic check       237.95  
4                 Yes  Bank transfer (automatic)               
...               ...                        ...          ...  
5629              Yes           Electronic check           95  
5630              Yes    Credit card (automatic)       2198.3  
5631              Yes           Electronic check       306.05  
5632              Yes           Electronic check      1200.15  
5633               No    Credit card (automatic)        457.3  

[5634 rows x 18 columns]

In [7]:
features_numeriche = dataset.select_dtypes(include = np.number)   #seleziono le features che sono numeri 
features_numeriche 

SeniorCitizen  tenure  MonthlyCharges
0                 0      21           64.85
1                 0      54           97.20
2                 0       1           23.45
3                 0       4           70.20
4                 0       0           61.90
...             ...     ...             ...
5629              0       1           95.00
5630              0      23           91.10
5631              0      12           21.15
5632              1      12           99.45
5633              0      26           19.80

[5634 rows x 3 columns]

In [8]:
# utilizzo il label encoder per trasformare i valori che non sono numerici in valori numerici
label_encoder = LabelEncoder()

for i in features_categoriche:
    features_categoriche[i] = label_encoder.fit_transform(features_categoriche[i])
    
features_categoriche

C:\Users\Erica\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


Churn  customerID  gender  Partner  Dependents  PhoneService  \
0         0        2370       0        0           1             1   
1         0        3457       0        0           0             1   
2         1        2149       1        1           0             0   
3         1        1426       1        0           0             1   
4         0        1545       1        0           1             1   
...     ...         ...     ...      ...         ...           ...   
5629      1         375       1        1           0             1   
5630      0        3418       0        1           1             1   
5631      0        4591       1        1           1             1   
5632      1        4247       1        0           0             1   
5633      0        2700       1        0           0             1   

      MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0                 0                0               2             0   
1                 2                1               0             2   
2                 1                0               0             0   
3                 0                1               0             0   
4                 2                0               2             2   
...             ...              ...             ...           ...   
5629              0                1               2             0   
5630              2                0               2             2   
5631              0                2               1             1   
5632              2                1               0             0   
5633              0                2               1             1   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                    2            0            0                2         1   
1                    0            0            2                2         2   
2                    0            0            0                0         0   
3                    0            0            0                0         0   
4                    0            2            0                0         2   
...                ...          ...          ...              ...       ...   
5629                 0            0            2                2         0   
5630                 2            2            2                2         2   
5631                 1            1            1                1         0   
5632                 2            0            2                2         0   
5633                 1            1            1                1         1   

      PaperlessBilling  PaymentMethod  TotalCharges  
0                    0              3           492  
1                    1              0          3511  
2                    1              2          1566  
3                    1              2          1627  
4                    1              0             0  
...                ...            ...           ...  
5629                 1              2          5231  
5630                 1              1          1466  
5631                 1              2          2167  
5632                 1              2           305  
5633                 0              1          3149  

[5634 rows x 18 columns]

Il labelEncoder trasforma i valori di tipo object in valori interi

In [9]:
newDataset = pd.concat([features_numeriche, features_categoriche], axis = 1) #concateno di nuovo le fatures numeriche con quelle trasformate

In [10]:
newDataset #mostro il nuovo dataset

SeniorCitizen  tenure  MonthlyCharges  Churn  customerID  gender  \
0                 0      21           64.85      0        2370       0   
1                 0      54           97.20      0        3457       0   
2                 0       1           23.45      1        2149       1   
3                 0       4           70.20      1        1426       1   
4                 0       0           61.90      0        1545       1   
...             ...     ...             ...    ...         ...     ...   
5629              0       1           95.00      1         375       1   
5630              0      23           91.10      0        3418       0   
5631              0      12           21.15      0        4591       1   
5632              1      12           99.45      1        4247       1   
5633              0      26           19.80      0        2700       1   

      Partner  Dependents  PhoneService  MultipleLines  ...  OnlineSecurity  \
0           0           1             1              0  ...               2   
1           0           0             1              2  ...               0   
2           1           0             0              1  ...               0   
3           0           0             1              0  ...               0   
4           0           1             1              2  ...               2   
...       ...         ...           ...            ...  ...             ...   
5629        1           0             1              0  ...               2   
5630        1           1             1              2  ...               2   
5631        1           1             1              0  ...               1   
5632        0           0             1              2  ...               0   
5633        0           0             1              0  ...               1   

      OnlineBackup  DeviceProtection  TechSupport  StreamingTV  \
0                0                 2            0            0   
1                2                 0            0            2   
2                0                 0            0            0   
3                0                 0            0            0   
4                2                 0            2            0   
...            ...               ...          ...          ...   
5629             0                 0            0            2   
5630             2                 2            2            2   
5631             1                 1            1            1   
5632             0                 2            0            2   
5633             1                 1            1            1   

      StreamingMovies  Contract  PaperlessBilling  PaymentMethod  TotalCharges  
0                   2         1                 0              3           492  
1                   2         2                 1              0          3511  
2                   0         0                 1              2          1566  
3                   0         0                 1              2          1627  
4                   0         2                 1              0             0  
...               ...       ...               ...            ...           ...  
5629                2         0                 1              2          5231  
5630                2         2                 1              1          1466  
5631                1         0                 1              2          2167  
5632                2         0                 1              2           305  
5633                1         1                 0              1          3149  

[5634 rows x 21 columns]

In [11]:
#dichiaro quale feature è X e quale è y
X = newDataset.drop(columns=['Churn']) #X sono tutte le colonne, quindi tutte le feature del cliente 
y = newDataset.Churn #y è la colonna target Churn

In [12]:
X

SeniorCitizen  tenure  MonthlyCharges  customerID  gender  Partner  \
0                 0      21           64.85        2370       0        0   
1                 0      54           97.20        3457       0        0   
2                 0       1           23.45        2149       1        1   
3                 0       4           70.20        1426       1        0   
4                 0       0           61.90        1545       1        0   
...             ...     ...             ...         ...     ...      ...   
5629              0       1           95.00         375       1        1   
5630              0      23           91.10        3418       0        1   
5631              0      12           21.15        4591       1        1   
5632              1      12           99.45        4247       1        0   
5633              0      26           19.80        2700       1        0   

      Dependents  PhoneService  MultipleLines  InternetService  \
0              1             1              0                0   
1              0             1              2                1   
2              0             0              1                0   
3              0             1              0                1   
4              1             1              2                0   
...          ...           ...            ...              ...   
5629           0             1              0                1   
5630           1             1              2                0   
5631           1             1              0                2   
5632           0             1              2                1   
5633           0             1              0                2   

      OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0                  2             0                 2            0   
1                  0             2                 0            0   
2                  0             0                 0            0   
3                  0             0                 0            0   
4                  2             2                 0            2   
...              ...           ...               ...          ...   
5629               2             0                 0            0   
5630               2             2                 2            2   
5631               1             1                 1            1   
5632               0             0                 2            0   
5633               1             1                 1            1   

      StreamingTV  StreamingMovies  Contract  PaperlessBilling  PaymentMethod  \
0               0                2         1                 0              3   
1               2                2         2                 1              0   
2               0                0         0                 1              2   
3               0                0         0                 1              2   
4               0                0         2                 1              0   
...           ...              ...       ...               ...            ...   
5629            2                2         0                 1              2   
5630            2                2         2                 1              1   
5631            1                1         0                 1              2   
5632            2                2         0                 1              2   
5633            1                1         1                 0              1   

      TotalCharges  
0              492  
1             3511  
2             1566  
3             1627  
4                0  
...            ...  
5629          5231  
5630          1466  
5631          2167  
5632           305  
5633          3149  

[5634 rows x 20 columns]

In [13]:
y

0       0
1       0
2       1
3       1
4       0
       ..
5629    1
5630    0
5631    0
5632    1
5633    0
Name: Churn, Length: 5634, dtype: int32

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.20,random_state=0)

Divido il train e il test set, con l'insieme di test al 20%

In [15]:
#decido di usare un classificatore DecisionTree
tree = DecisionTreeClassifier(criterion='entropy',random_state=1,max_depth=6)
tree.fit(X_train,y_train)
predictions = tree.predict(X_test)

In [16]:
print("Accuracy : "+str(accuracy_score(y_test,predictions)))

Accuracy : 0.7888198757763976


Con il DecisionTree raggiungo un'accuratezza del 78,8%

In [17]:
confusion_matrix = confusion_matrix(y_test, predictions)
print(confusion_matrix)

[[764  78]
 [160 125]]


La matrice di confusione mostra 764 e 125 sono previsioni giuste e 78 e 160 sono previsioni errate

In [18]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.83      0.91      0.87       842
           1       0.62      0.44      0.51       285

    accuracy                           0.79      1127
   macro avg       0.72      0.67      0.69      1127
weighted avg       0.77      0.79      0.78      1127



Provo ora ad utilizzare una random forest per vedere se aumenta l'accuratezza

### La foresta casuale (o Random Forest) è un tipo di algoritmo di apprendimento automatico supervisionato basato sull’apprendimento di più modelli previsionali per formare un unico modello di previsione più potente.Ogni modello utilizzato dalla previsione del Random Forest di solito è un albero decisionale.
In questo caso decido di utilizzare 100 alberi decisionali

In [19]:
from sklearn.ensemble import RandomForestClassifier

In [30]:
random_classifier = RandomForestClassifier(criterion='entropy',n_estimators=100, random_state=1)
random_classifier.fit(X_train,y_train)
y_pred = random_classifier.predict(X_test)

In [31]:
print("Accuracy : "+str(accuracy_score(y_test,y_pred)))

Accuracy : 0.8003549245785271


L'accuratezza sale all'80%

In [29]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.83      0.92      0.87       842
           1       0.65      0.46      0.54       285

    accuracy                           0.80      1127
   macro avg       0.74      0.69      0.71      1127
weighted avg       0.79      0.80      0.79      1127

